godot-core/src/private.rs

```diff
@@ -351,19 +350,285 @@ impl ScopedFunctionStack {
     }
 }
 
-#[cfg(all(debug_assertions, not(wasm_nothreads)))]
-thread_local! {
+/// A thread local which adequately behaves as a global variable when compiling under `experimental-wasm-nothreads`, as it does
+/// not support thread locals. Aims to support similar APIs as [`std::thread::LocalKey`].
+pub(crate) struct GodotThreadLocal<T: 'static> {
+    #[cfg(not(wasm_nothreads))]
+    threaded_val: &'static std::thread::LocalKey<T>,
+
+    #[cfg(wasm_nothreads)]
+    non_threaded_val: std::cell::OnceCell<T>,
+
+    #[cfg(wasm_nothreads)]
+    initializer: fn() -> T,
+}
+
+// SAFETY: there can only be one thread with `wasm_nothreads`.
+#[cfg(wasm_nothreads)]
+unsafe impl<T: 'static> Sync for GodotThreadLocal<T> {}
+
+impl<T: 'static> GodotThreadLocal<T> {
+    #[cfg(not(wasm_nothreads))]
+    pub const fn new_threads(key: &'static std::thread::LocalKey<T>) -> Self {
+        Self { threaded_val: key }
+    }
+
+    #[cfg(wasm_nothreads)]
+    pub const fn new_nothreads(initializer: fn() -> T) -> Self {
+        Self {
+            non_threaded_val: std::cell::OnceCell::new(),
+            initializer,
+        }
+    }
+
+    /// Acquires a reference to the value in this TLS key.
+    ///
+    /// See [`std::thread::LocalKey::with`] for details.
+    pub fn with<F, R>(&'static self, f: F) -> R
+    where
+        F: FnOnce(&T) -> R,
+    {
+        #[cfg(not(wasm_nothreads))]
+        return self.threaded_val.with(f);
+
+        #[cfg(wasm_nothreads)]
+        f(self.non_threaded_val.get_or_init(self.initializer))
+    }
+
+    /// Acquires a reference to the value in this TLS key.
+    ///
+    /// See [`std::thread::LocalKey::try_with`] for details.
+    #[allow(dead_code)]
+    #[inline]
+    pub fn try_with<F, R>(&'static self, f: F) -> Result<R, std::thread::AccessError>
+    where
+        F: FnOnce(&T) -> R,
+    {
+        #[cfg(not(wasm_nothreads))]
+        return self.threaded_val.try_with(f);
+
+        #[cfg(wasm_nothreads)]
+        Ok(self.with(f))
+    }
+}
+
+#[allow(dead_code)]
+impl<T: 'static> GodotThreadLocal<Cell<T>> {
+    /// Sets or initializes the contained value.
+    ///
+    /// See [`std::thread::LocalKey::set`] for details.
+    pub fn set(&'static self, value: T) {
+        #[cfg(not(wasm_nothreads))]
+        return self.threaded_val.set(value);
+
+        // According to `LocalKey` docs, this method must not call the default initializer.
+        #[cfg(wasm_nothreads)]
+        if let Some(initialized) = self.non_threaded_val.get() {
+            initialized.set(value);
+        } else {
+            self.non_threaded_val.get_or_init(|| Cell::new(value));
+        }
+    }
+
+    /// Returns a copy of the contained value.
+    ///
+    /// See [`std::thread::LocalKey::get`] for details.
+    pub fn get(&'static self) -> T
+    where
+        T: Copy,
+    {
+        #[cfg(not(wasm_nothreads))]
+        return self.threaded_val.get();
+
+        #[cfg(wasm_nothreads)]
+        self.with(Cell::get)
```

Comment:
This differentiation occurs a dozen times, is it not possible to abstract?

Something like
```rs
self.value().get()
```
where `value()` is `#[cfg]`ed?

Or is the `with` essential in all methods?

Reply:
I'm using `with` even for `nothreads` to keep the default initialization logic confined to `with`, for convenience (I used to have a separate method like the one you propose, but it felt like a waste of space). Also, this let me simply copy and paste the important part of the implementation of each method from stdlib, giving me confidence that I'm not messing it up. :p

Reply:
Sorry I might have misunderstood - if you mean that we could replace all cfg calls with `.value()`, then no, as LocalKey imposes the usage of `.with` with a callback to read the value.

Some of the cfgs are not really required (particularly some on the `RefCell` and `Cell` helpers), but I felt like keeping them in case some implementation detail on their side changes in a future rust version.